src/store.cc

```diff
@@ -149,8 +149,10 @@ StoreEntry::makePublic(const KeyScope scope)
 void
 StoreEntry::makePrivate(const bool shareable)
 {
-    /* This object should never be cached at all */
-    expireNow();
+    if (!shareable) {
+        // This object should not be reused
+        expireNow();
```

Comment:
Why do we need to expire entries when making them private?

Reply:
This makePrivate(false) is called when the decision is 'reuseNot': we can't share the entry even with already collapsed clients. E.g.,  we got a header forbidding any caching/sharing (smth. like CC:no-store).

Reply:
Why do we need to expire entries when making them private and unshareable?

Reply:
Probably this sounds vague, but to my mind 'expiration' prevents sharing shareable but not reusable entries. I.e., it makes difference between doNotCacheButShare and reuseNot states.
I am not insist this is [the only] correct way for doing this. This is how it is implemented and working now.

Reply:
> 'expiration' prevents sharing shareable but not reusable entries

True, but should not be relevant because we are talking about unshareable entries.

> I.e., it makes difference between doNotCacheButShare and reuseNot states.

My argument that it should _not_ make a difference. I do not know whether it does. Unshareable entries should be unshareable regardless of their expiration time. If unsharable entries are shareable, there is a bug we should fix.

> This is how it is implemented and working now.

I know but that is largely irrelevant to my question. If the current code works correctly (in some cases) because of a bad hack, and we are fixing a very related aspect of that code because of a very related bug (in other cases), then we should remove the hack instead of splitting the code into two parts, one that continues to rely on a hack and one that is no longer affected by the bugs that hack caused.

If you remove that if statement and expireNow() call, will the code break in some cases?

* If yes, then the code relies on a weak expiration check to prevent sharing of unsharable entries. Such code must be fixed IMO.
* If no, then the code is already OK and the expireNow() call is simply not needed (after all your other changes/fixes).

Reply:
> If no, then the code is already OK and the expireNow() call is simply not needed (after all your other changes/fixes).

I think you are right with this. I overlooked  our StoreEntry::mayStartHitting(), developed to distinguish such shareable and unshareable entries. So there should be no need to do extra work with expireNow().

> will the code break in some cases

It is difficult to say. I can state only that the major code paths (clientReplyContext::cacheHit() or clientReplyContext::handleIMSReply()) already use mayStartHitting() and there is no need there for protection with 'expiration'.